codec/src/main/java/io/netty/handler/codec/compression/ZstdEncoder.java

```diff
@@ -0,0 +1,326 @@
+        }
+
+        if (preferDirect) {
+            return ctx.alloc().ioBuffer(targetBufSize, targetBufSize);
+        } else {
+            return ctx.alloc().heapBuffer(targetBufSize, targetBufSize);
+        }
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws Exception {
+        if (finished) {
+            if (!out.isWritable(in.readableBytes())) {
+                // out should be EMPTY_BUFFER because we should have allocated enough space above in allocateBuffer.
+                throw ENCODE_FINSHED_EXCEPTION;
+            }
+            out.writeBytes(in);
+            return;
+        }
+
+        final ByteBuf buffer = this.buffer;
+        int length;
+        while ((length = in.readableBytes()) > 0) {
+            final int nextChunkSize = Math.min(length, buffer.writableBytes());
+            in.readBytes(buffer, nextChunkSize);
+
+            if (nextChunkSize <= MIN_BLOCK_SIZE) {
+                writeUnCompressedData(out);
+                return;
+            }
+
+            if (!buffer.isWritable()) {
+                flushBufferedData(out);
+            }
+        }
+        if (buffer.isReadable()) {
+            flushBufferedData(out);
+        }
+    }
+
+    private void flushBufferedData(ByteBuf out) {
+        final int flushableBytes = buffer.readableBytes();
+        if (flushableBytes == 0) {
+            return;
+        }
+        checksum.reset();
+        checksum.update(buffer.internalNioBuffer(buffer.readerIndex(), flushableBytes));
+        final int check = (int) checksum.getValue();
+
+        final int bufSize = (int) Zstd.compressBound(flushableBytes) + HEADER_LENGTH;
+        out.ensureWritable(bufSize);
+        final int idx = out.writerIndex();
+        int compressedLength;
+        try {
+            ByteBuffer outNioBuffer = out.internalNioBuffer(idx + HEADER_LENGTH, out.writableBytes() - HEADER_LENGTH);
+            compressedLength = Zstd.compress(
+                    outNioBuffer,
+                    buffer.internalNioBuffer(buffer.readerIndex(), flushableBytes),
+                    DEFAULT_COMPRESS_LEVEL);
+        } catch (Exception e) {
+            throw new CompressionException(e);
+        }
+        final int blockType;
+        if (compressedLength >= flushableBytes) {
+            blockType = BLOCK_TYPE_NON_COMPRESSED;
+            compressedLength = flushableBytes;
+            out.setBytes(idx + HEADER_LENGTH, buffer, 0, flushableBytes);
+        } else {
+            blockType = BLOCK_TYPE_COMPRESSED;
+        }
+
+        out.setInt(idx, ZSTD_NUMBER);
+        out.setByte(idx + TOKEN_OFFSET, (byte) (blockType | compressionLevel));
+        out.setIntLE(idx + COMPRESSED_LENGTH_OFFSET, compressedLength);
+        out.setIntLE(idx + DECOMPRESSED_LENGTH_OFFSET, flushableBytes);
+        out.setIntLE(idx + DEFAULT_CHECKSUM_OFFSET, check);
+        out.writerIndex(idx + HEADER_LENGTH + compressedLength);
```

Comment:
Does this framing or metadata follow some standard, or is it specific to this implementation in Netty?

Reply:
This is a specific custom protocol, not an implementation that follows a certain framework